cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -205,150 +256,211 @@ std::unique_ptr<column> generate_entry_list_offsets(size_type num_entries,
 }
 
 /**
- * @brief Performs an equality comparison between two entries in a lists column
+ * @brief Performs an equality comparison between two entries in a lists column.
  *
- * For the two elements that are in the same list in the lists column, they will always be
- * considered as different. If they are from the same list and their type is one of floating
- * point types, this functor will return the same comparison result as
- * `cudf::element_equality_comparator`.
+ * For the two elements that are NOT in the same list in the lists column, they will always be
+ * considered as different. If they are from the same list and their type is not floating point,
+ * this functor will return the same comparison result as `cudf::element_equality_comparator`.
  *
  * For floating-point types, entries holding NaN value can be considered as different values or the
- * same value depending on the nans_equal parameter.
+ * same value depending on the `nans_equal` parameter.
  *
- * @tparam Type       The data type of entries
+ * @tparam Type The data type of entries
  * @tparam nans_equal Flag to specify whether NaN entries should be considered as equal value (only
  * applicable for floating-point data column)
  */
-template <class Type, bool nans_equal>
-class list_entry_comparator {
- public:
-  list_entry_comparator(offset_type const* list_offsets,
-                        column_device_view d_view,
-                        null_equality nulls_equal,
-                        bool has_nulls)
-    : list_offsets(list_offsets), d_view{d_view}, nulls_equal{nulls_equal}, has_nulls(has_nulls)
+template <class Type>
+struct column_row_comparator_fn {
+  template <typename T, std::enable_if_t<!cuda::std::is_floating_point_v<T>>* = nullptr>
+  bool __device__ compare(T const& lhs_val, T const& rhs_val) const noexcept
   {
+    return lhs_val == rhs_val;
   }
 
-  template <bool nans_equal_ = nans_equal>
-  std::enable_if_t<cuda::std::is_floating_point_v<Type> and nans_equal_, bool> __device__
-  operator()(size_type i, size_type j) const noexcept
+  template <typename T, std::enable_if_t<cuda::std::is_floating_point_v<T>>* = nullptr>
+  bool __device__ compare(T const& lhs_val, T const& rhs_val) const noexcept
   {
-    // Two entries are not considered for equality if they belong to different lists
+    // If both element(i) and element(j) are NaNs and nans are considered as equal value then this
+    // comparison will return `true`. This is the desired behavior in Pandas.
```

Comment:
Maybe this is something we can improve in a future PR if it's too big a change now, but do we want to embed pandas/Spark specific behavior this deep in the call stack rather than having some configuration value passed down? I can see that it would need to get passed through about 4 levels of calls to get here, which would be annoying, but it also seems more in line with our general design philosophy to have an enum to configure this behavior rather than encoding it here.

Reply:
Absolutely agree with this. When this API was first reviewed, there was a long discussion about how to satisfy both Pandas' users and Spark's users, since each side requires a different behavior (NaN are equal or not). In the meantime, I don't have a better idea how to avoid this deep call stack. I'm happy to do it (or anybody can do it) if there is a better way.

Reply:
Maybe I'm missing something, but would it not be possible to define an enum for different nan behaviors and then pass that down (all the way through `get_unique_entries_and_list_offsets`->`get_unique_entries_dispatch`... I know it's a lot of forwarding) to control this behavior based on a templated version of the `column_row_comparator_fn`? I'm not saying that we need to do it in this PR, but if that solution would work and be preferable we should at least make an issue for this to avoid losing track once the PR is merged.

Reply:
Thanks. I'll take a note on this. There should be a room for improvement.

Reply:
Done. Here it is: https://github.com/rapidsai/cudf/issues/9257